File: ssotk/vocab.py
NAME_CHARS = set(
    b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789/\\._- "
)


class KH:
    OFFSET      = 0x0B6EF964
    SIZE        = 0x0B1A6E08
    NAME_OFFSET = 0x01C65A88
    EXTRACT     = 0x0E89DE15
    OWN         = 0x06D029F4
    PARENT      = 0x0FA8A5A4
    ID          = 0x040034D4
    BREED_ID    = 0x0BB21504
    PRICE       = 0x01DA6705
    JS_PRICE    = 0x088D4FB5
    SC_PRICE    = 0x008D4FA5
    LEVEL       = 0x06A98294


HASH_NAMES: dict[int, str] = {
    KH.OFFSET:      "offset",
    KH.SIZE:        "size",
    KH.NAME_OFFSET: "name_offset",
    KH.EXTRACT:     "extract",
    KH.OWN:         "own",
    KH.PARENT:      "parent",
    KH.ID:          "id",
    KH.BREED_ID:    "breed_id",
    KH.PRICE:       "price",
    KH.JS_PRICE:    "price_js",
    KH.SC_PRICE:    "price_sc",
    KH.LEVEL:       "level",
}


def deobfuscate(data: bytes, key: int) -> bytes:
    return bytes((b + key) & 0xFF for b in data)


def _looks_text(b: bytes) -> bool:
    return bool(b) and all(c in NAME_CHARS for c in b)


def auto_deobfuscate(data: bytes) -> tuple[bytes, int | None]:
    if _looks_text(data):
        return data, 0
    for key in range(1, 256):
        if data and all(((b + key) & 0xFF) in NAME_CHARS for b in data):
            return deobfuscate(data, key), key
    return data, None


def name_for_hash(h: int) -> str:
    return HASH_NAMES.get(h, f"0x{h:08x}")


try:
    from . import _core as _c

    def deobfuscate(data: bytes, key: int) -> bytes:  # noqa: F811
        return _c.vocab.deobfuscate(data, key)

    def auto_deobfuscate(data: bytes) -> tuple[bytes, int | None]:  # noqa: F811
        return _c.vocab.auto_deobfuscate(data)

    def name_for_hash(h: int) -> str:  # noqa: F811
        return _c.vocab.name_for_hash(h)
except ImportError:
    pass

File: native/core/src/vocab.cpp
#include "ssotk/vocab.hpp"

#include <array>
#include <cstdio>

namespace ssotk::vocab {

namespace {

constexpr std::array<bool, 256> make_name_charset() {
    std::array<bool, 256> t{};
    for (int c = 'a'; c <= 'z'; ++c) t[c] = true;
    for (int c = 'A'; c <= 'Z'; ++c) t[c] = true;
    for (int c = '0'; c <= '9'; ++c) t[c] = true;
    for (unsigned char c : std::string_view("/\\._- ")) t[c] = true;
    return t;
}

constexpr auto NAME_CHARS = make_name_charset();

bool looks_text(std::string_view b) {
    if (b.empty()) return false;
    for (unsigned char c : b) {
        if (!NAME_CHARS[c]) return false;
    }
    return true;
}

}

std::string deobfuscate(std::string_view in, uint8_t key) {
    std::string out(in.size(), '\0');
    for (size_t i = 0; i < in.size(); ++i) {
        out[i] = static_cast<char>((static_cast<uint8_t>(in[i]) + key) & 0xFF);
    }
    return out;
}

std::pair<std::string, std::optional<uint8_t>> auto_deobfuscate(std::string_view data) {
    if (looks_text(data)) return {std::string(data), uint8_t{0}};
    if (data.empty())     return {std::string(data), std::nullopt};
    for (int key = 1; key < 256; ++key) {
        bool ok = true;
        for (unsigned char c : data) {
            if (!NAME_CHARS[static_cast<uint8_t>((c + key) & 0xFF)]) { ok = false; break; }
        }
        if (ok) return {deobfuscate(data, static_cast<uint8_t>(key)), static_cast<uint8_t>(key)};
    }
    return {std::string(data), std::nullopt};
}

const std::unordered_map<uint32_t, std::string>& hash_names() {
    static const std::unordered_map<uint32_t, std::string> table = {
        {KH::OFFSET,      "offset"},
        {KH::SIZE,        "size"},
        {KH::NAME_OFFSET, "name_offset"},
        {KH::EXTRACT,     "extract"},
        {KH::OWN,         "own"},
        {KH::PARENT,      "parent"},
        {KH::ID,          "id"},
        {KH::BREED_ID,    "breed_id"},
        {KH::PRICE,       "price"},
        {KH::JS_PRICE,    "price_js"},
        {KH::SC_PRICE,    "price_sc"},
        {KH::LEVEL,       "level"},
    };
    return table;
}

std::string name_for_hash(uint32_t h) {
    const auto& t = hash_names();
    auto it = t.find(h);
    if (it != t.end()) return it->second;
    char buf[16];
    std::snprintf(buf, sizeof(buf), "0x%08x", h);
    return std::string(buf);
}

}
